Label backgrounds

emackey opened the pull request:
This PR adds three options to `Label` primitives and `LabelGraphics` for entities & CZML.

* `showBackground` (default: `false`)
* `backgroundColor` (default: `new Cesium.Color(0.165, 0.165, 0.165, 0.8)`, translucent dark gray)
* `backgroundPadding` (default: `new Cesium.Cartesian2(2, 2)` but this might be a bit small)

Additionally, this PR adds a new value to `VerticalOrigin` called `BASELINE`, which for Billboards is just a synonym for `BOTTOM`, but for labels will align the text along the baseline (for example with lowercase `y` hanging down below the baseline).  (Note that master's behavior is somewhat unpredictable now, depending on heights and descenders of glyphs, master's `BOTTOM` sometimes acts as a baseline and sometimes as a true bottom.  For example try label `oy` vs `Oy` in master and look at the bottom alignment).

~~I'm getting this open for review now, to get on the agenda for the code sprint, but it's not quite ready to merge yet.  The main blocker is there appears to be a crash bug when terrain, `HeightReference`, and `showBackground` are used together *and then removed* from the viewer.  For an example of this, load the Ground Clamping sandcastle demo on this branch, and use the dropdown to switch to any other clamped entity choice.  The point + label demo works, but trashes the frustum list when the entity is removed.  This is tough to track down and any assistance would be appreciated.~~

**TODO**

- [x] Fix bug revealed by Ground Clamping demo. -- was caused by rendering empty text label with background turned on.
- [x] Test behavior of backgrounds when mixed with clustering. -- fixed typo that affected this.
- [x] Post demo I wrote for visually testing `getScreenSpacePosition` to justify changes there.
- [x] Add backgrounds and `BASELINE` to various label/billboard demos.
- [x] CHANGES.md

Separately:

- [ ] Submit PR to `czml-writer`.

pjcozzi commented:
On Mac/Chrome (didn't test anything else), the background for the label in the Ground Clamping Sandcastle example jumps horizontally when the camera moves.

![ok](https://cloud.githubusercontent.com/assets/782098/20902925/5de45fae-bb07-11e6-8f26-6bf10e0783f0.gif)

pjcozzi commented:
Can you add a few examples to the Label Sandcastle example or add this to the above tasklist?

pjcozzi commented:
> backgroundPadding (default: new Cesium.Cartesian2(2, 2) but this might be a bit small)

I don't know if this has anything to do with the Retina display, but I agree it looks small.  In particular, look at the left and top of the "P" - 

![image](https://cloud.githubusercontent.com/assets/782098/20903060/f2211464-bb07-11e6-8271-1d8eaf25e3be.png)

pjcozzi commented on Source/Scene/Label.js:
```diff
@@ -337,6 +353,82 @@ define([
         /**
+         * Determines if a background behind this label will be shown.
+         * @memberof Label.prototype
+         * @type {Boolean}
```
Here and below, add `@default`.

pjcozzi commented on Source/Scene/Label.js:
```diff
@@ -337,6 +353,82 @@ define([
+                var backgroundPadding = this._backgroundPadding;
+                if (!Cartesian2.equals(backgroundPadding, value)) {
+                    Cartesian2.clone(value, backgroundPadding);
+                    repositionAllGlyphs(this);
```
Should this call to `repositionAllGlyphs` and the above call to `rebindAllGlyphs` in `showBackground` (and perhaps others) be delayed until the next `update`?  If they are cheap, just leave them.  If they are expensive and potentially redundant in common code like:
```javascript
label.showBackground = true; // does something expensive
label.backgroundPadding = /* ... */ // does something redundant and expensive again
```
consider delaying them.

emackey replied:
I believe [this delay already happens](https://github.com/AnalyticalGraphicsInc/cesium/blob/178d71253977472bcc1934705e26151e453da939/Source/Scene/Label.js#L36-L50).

pjcozzi commented on Source/Scene/Label.js:
```diff
@@ -825,38 +957,58 @@ define([
      * @private
      */
     Label.getScreenSpaceBoundingBox = function(label, screenSpacePosition, result) {
```
I haven't looked at the tests yet, but do we have enough tests to cover the new cases here?

emackey replied:
I think so.  I worked on improving the code coverage for labels, the new code and some previously untested conditionals within older code are covered now.

pjcozzi commented:
Throughout, do you think the reference doc is clear enough on the relationship of the label's position/anchor point and the background?  Do any other parts need to be updated?

pjcozzi commented on Source/Scene/LabelCollection.js:
```diff
@@ -339,13 +437,21 @@ define([
         this._scene = options.scene;
 
         this._textureAtlas = undefined;
+        this.__backgroundTextureAtlas = undefined;
```
Do you mean `_`, not `__`?

emackey replied:
Good catch

pjcozzi commented:
Added unit tests to the above tasklist.

pjcozzi commented on Source/Scene/LabelCollection.js:
```diff
@@ -339,13 +437,21 @@ define([
         this._billboardCollection.destroyTextureAtlas = false;
 
         this._spareBillboards = [];
+        this._spareBackgroundBillboards = [];
```
Are you sure about needing this?  I don't think we optimize for cases like this anywhere else in Cesium.

pjcozzi replied:
I'm asking if we need the spares at all.

emackey replied:
Good question.  I copied the spare glyph logic without re-thinking the reasoning why spares exist in the first place.  Does the Billboard system do a good enough job internally, destroying and creating new billboards, that it's not worth holding onto any spare billboards at all?  I'll see if I can investigate this a bit myself, but would appreciate any input from someone familiar with how billboards are destroyed and created in a collection.

pjcozzi replied:
Add/remove billboards causes the GL buffer to be recreated in the next `update` (so that is per frame, not per billboard).

The spares appear to be an "optimization" for setting show to true/false/true.  Probably not needed.

pjcozzi replied:
Sounds good.

pjcozzi commented:
> Added unit tests to the above tasklist.

Nevermind, the diff wasn't shown by default because there are so many new tests!

pjcozzi commented on Specs/Scene/LabelCollectionSpec.js:
```diff
@@ -875,16 +948,23 @@ defineSuite([
 
-                width += billboard.width;
-                height = Math.max(height, billboard.height);
+                var glyphWidth = billboard.width * scale;
```
This code is duplicate a lot; it is worth adding a utility function to this file?

emackey replied:
Yeah, to be honest I'm not happy with this repeated code being here at all.  Different test platforms have different font metrics, so the "correct" answers need to be calculated inside the test.  As a result these tests (in master) have a copy-and-paste of part of the body of the function they are testing in the test itself.  I'm no testing expert but I'm sure that violates some kind of cardinal rule of unit tests.  Essentially it calls the function to test, and calls a local copy-and-paste of the meat of that same function, and expects the results to match, and yes they do.

I'd like to replace this with something, but it needs to work across platforms.  Maybe we need a way to mock the Cesium `writeTextToCanvas` or the 3rd-party `measureText` for the purpose of testing just the screen space calculations.

pjcozzi replied:
Is it possible to write a more coarse-grained expectation so we don't have to duplicate parts of the implementation in the tests?  Even if the expectation isn't as precise, it is probably more useful and robust across platforms.

pjcozzi commented:
@emackey looks pretty solid, just those comments.

> I'm getting this open for review now, to get on the agenda for the code sprint, but it's not quite ready to merge yet. The main blocker is there appears to be a crash bug when terrain, HeightReference, and showBackground are used together and then removed from the viewer. For an example of this, load the Ground Clamping sandcastle demo on this branch, and use the dropdown to switch to any other clamped entity choice. The point + label demo works, but trashes the frustum list when the entity is removed. This is tough to track down and any assistance would be appreciated.

I can reproduce this, but I don't have any ideas off the top of my head other than the obvious `_backgroundBillboardCollection` is not playing fairly / timing of the terrain height updates.  @bagnell may have a more precise idea.

emackey commented on Source/Scene/LabelCollection.js:
```diff
@@ -103,6 +125,44 @@ define([
 
+        var showBackground = label._showBackground;
+        var backgroundBillboard = label._backgroundBillboard;
+        if (!showBackground) {
```
Possibly, I'm just mirroring what is already done with `_spareBillboards` for glyphs.  We can't have the spare billboards (for glyphs or backgrounds) shared between separate label collections, which is probably why the spares are owned by the collection instance.

pjcozzi replied:
I'm suggesting a fairly trivial change:
```javascript
var spareBackgroundBillboards = labelCollection._spareBackgroundBillboards;
// use spareBackgroundBillboards below
```
Just a style thing from the Coding Guide: https://github.com/AnalyticalGraphicsInc/cesium/tree/master/Documentation/Contributors/CodingGuide#basic-code-construction

emackey replied:
Oh, I misunderstood.

emackey commented:
For reference, here's the code I'm using to visually inspect the results of private function `getScreenSpaceBoundingBox`.  I was tempted to make this function public, unless there's a reason not to.

This code draws a pink DOM border around the reported position.

```
var viewer = new Cesium.Viewer('cesiumContainer');
//viewer.cesiumWidget.resolutionScale = 0.25;

var scene = viewer.scene;
var position = Cesium.Cartesian3.fromDegrees(-75.1641667, 39.9522222);

var pointPrimitives = scene.primitives.add(new Cesium.PointPrimitiveCollection({scene: scene}));
var pointPrimitie1 = pointPrimitives.add({
    pixelSize : 5,
    color : Cesium.Color.YELLOW,
    position : position
});

var labels = viewer.scene.primitives.add(new Cesium.LabelCollection({scene: scene}));
var label1 = labels.add({
    position : position,
    text : 'Philly',
    font : '24px Helvetica',
    fillColor : new Cesium.Color(0.6, 0.9, 1.0),
    style : Cesium.LabelStyle.FILL,
    horizontalOrigin : Cesium.HorizontalOrigin.LEFT,
    verticalOrigin : Cesium.VerticalOrigin.TOP,
    //pixelOffset : new Cesium.Cartesian2(100, 50),
    scale : 1.5,
    showBackground : true
});


var div = document.createElement('div');
div.style.border = '1px solid #f0f';
div.style.position = 'absolute';
document.body.appendChild(div);

var scratch1 = new Cesium.Cartesian3();
var rect = new Cesium.BoundingRectangle();
viewer.clock.onTick.addEventListener(function() {
    var screenSpacePosition = label1.computeScreenSpacePosition(viewer.scene, scratch1);
    if (!Cesium.defined(screenSpacePosition)) {
        div.style.top = '-100px';
        div.style.height = '0';
    } else {
        Cesium.Label.getScreenSpaceBoundingBox(label1, screenSpacePosition, rect);
        div.style.left = rect.x - 1 + 'px';
        div.style.top = rect.y - 1 + 'px';
        div.style.width = rect.width + 'px';
        div.style.height = rect.height + 'px';
    }
});

viewer.scene.fxaa = false;

Sandcastle.addToolbarMenu([{
    text : 'HorizontalOrigin.LEFT',
    onselect : function() {
        label1.horizontalOrigin = Cesium.HorizontalOrigin.LEFT;
    }
}, {
    text : 'HorizontalOrigin.CENTER',
    onselect : function() {
        label1.horizontalOrigin = Cesium.HorizontalOrigin.CENTER;
    }
}, {
    text : 'HorizontalOrigin.RIGHT',
    onselect : function() {
        label1.horizontalOrigin = Cesium.HorizontalOrigin.RIGHT;
    }
}]);

Sandcastle.addToolbarMenu([{
    text : 'VerticalOrigin.TOP',
    onselect : function() {
        label1.verticalOrigin = Cesium.VerticalOrigin.TOP;
    }
}, {
    text : 'VerticalOrigin.CENTER',
    onselect : function() {
        label1.verticalOrigin = Cesium.VerticalOrigin.CENTER;
    }
}, {
    text : 'VerticalOrigin.BASELINE',
    onselect : function() {
        label1.verticalOrigin = Cesium.VerticalOrigin.BASELINE;
    }
}, {
    text : 'VerticalOrigin.BOTTOM',
    onselect : function() {
        label1.verticalOrigin = Cesium.VerticalOrigin.BOTTOM;
    }
}]);

Sandcastle.addToolbarMenu([{
    text : 'Show background',
    onselect : function() {
        label1.showBackground = true;
    }
}, {
    text : 'Hide background',
    onselect : function() {
        label1.showBackground = false;
    }
}]);


window.Cesium = Cesium;
window.viewer = viewer;
window.myDiv = div;

```

emackey commented:
This is almost ready.  I still need to investigate @pjcozzi's feedback on a better way to unit-test the getScreenSpacePosition function.  Also I plan to make a separate branch where I incorporate the multi-line logic from #4306, but I won't get to that today.

pjcozzi commented:
Will review again when this is fully ready unless there is something specific you want looked at now.

emackey commented:
It looks like the 1-pixel "shake" mentioned at the top of this review, in the Ground Clamping demo, is something that happens subtly to all billboards and glyphs when clamping is active, possibly due to small rounding errors.  The sharp edge of the background billboard can make this much more noticeable than on other billboards, drawing attention to itself.  But I think this is an unrelated existing issue with clamped billboards.

emackey commented:
I believe this is ready!

pjcozzi commented:
> It looks like the 1-pixel "shake" mentioned at the top of this review, in the Ground Clamping demo, is something that happens subtly to all billboards and glyphs...

Can you submit a separate issue now that this is more prominent?

pjcozzi commented:
@emackey can you run all the tests in Chrome and Firefox?

On Mac/Chrome, I have this new test failure:

> Scene/LabelCollection Label computes screen space bounding box with vertical origin center
Expected -18 to be less than -18.

pjcozzi commented:
@emackey did you test this with the various pixel and eye offset options?  Do these need more unit tests?

pjcozzi commented:
Just those comments.  I'll be online for the next few hours to merge if this is ready.

emackey commented:
That vertical center test got bad numbers in it with my last update.  Fixed.

The unit tests don't currently test combinations of features, but I did test all the combinations with backgrounds I could think of with hand-edits to the above pasted Sandcastle demo.  Generally these "just work" because the background itself is a billboard just like each glyph, and the various label offsets and options are merely copied down into the supporting billboards.  The clustering options were tested later in the review and revealed a typo there, now fixed.  All of the other options appear to be working correctly.

This is ready.

pjcozzi commented:
Awesome work, @emackey!

pjcozzi commented:
@emackey if any, please close any related issues or reply to any related forum threads.

shunter commented on Source/DataSources/LabelGraphics.js:
```diff
@@ -35,6 +35,9 @@ define([
      * @param {Property} [options.outlineColor=Color.BLACK] A Property specifying the outline {@link Color}.
      * @param {Property} [options.outlineWidth=1.0] A numeric Property specifying the outline width.
      * @param {Property} [options.show=true] A boolean Property specifying the visibility of the label.
+     * @param {Property} [options.showBackground=true] A boolean Property specifying the visibility of the background behind the label.
```
Wrong default is specified here, FYI.

emackey replied:
Thanks!
